ttyplot.c

```diff
@@ -289,9 +290,99 @@ void redraw_screen(const char * errstr) {
     refresh();
 }
 
+// Handle a single value from the input stream.
+// Return whether we got a full data record.
+int handle_value(double value)
+{
+    static double saved_value;
+    static int saved_value_valid = 0;
+
+    // First value of a 2-value record: save it for later.
+    if (two && !saved_value_valid) {
+        saved_value = value;
+        saved_value_valid = 1;
+        return 0;
+    }
+
+    // Otherwise we have a full record.
+    n = (n+1) % plotwidth;
+    if (two) {
+        values1[n] = saved_value;
+        values2[n] = value;
+        saved_value_valid = 0;
+    } else {
+        values1[n] = value;
+    }
+    if (rate)
+        td = derivative(&values1[n], two ? &values2[n] : NULL, &now);
+    return 1;
+}
+
+// Parse the input data. Return the number of bytes consumed.
+size_t parse_input(char *buffer, size_t length)
+{
+    static const char delimiters[] = " \t\r\n";  // white space
+
+    // Find the last delimiter.
+    int end;
+    for (end = length - 1; end >= 0; end--) {
+        if (strchr(delimiters, buffer[end]))  // found
+            break;
+    }
+    if (end < 0)  // no delimiter
+        return 0;
+    buffer[end] = '\0';
+
+    // Tokenize and parse.
+    int records = 0;  // number or records found
+    char *str = buffer;
+    char *token;
+    while ((token = strtok(str, delimiters)) != NULL) {
+        records += handle_value(strtod(token, NULL));
+        str = NULL;  // tell strtok() to stay on the same string
+    }
+    v += records;
+    if (records > 0)
+        redraw_screen(errstr);
+    return end + 1;
+}
+
+// Handle an "input ready" event, where only a single read() is guaranteed to not block.
+// Return whether the input stream got closed.
+int handle_input(void)
+{
+    static char buffer[4096];
+    static size_t buffer_pos = 0;
+
+    // Buffer incoming data.
+    int bytes_read = read(STDIN_FILENO, buffer + buffer_pos, sizeof buffer - 1 - buffer_pos);
+    if (bytes_read < 0) {  // read error
+        if (errno == EINTR || errno == EAGAIN)  // we should try again later
+            return 0;
+        errstr = strerror(errno);  // other errors are considered fatal
+        return 1;
+    }
+    if (bytes_read == 0) {
+        errstr = "input stream closed";
+        buffer[buffer_pos++] = '\n';  // attempt to extract one last value
+        parse_input(buffer, buffer_pos);
+        redraw_screen(errstr);  // redraw to display the error message
+        return 1;
+    }
+    buffer_pos += bytes_read;
+
+    // Handle this new data.
+    size_t bytes_consumed = parse_input(buffer, buffer_pos);
+    if (buffer_pos - bytes_consumed > sizeof buffer / 2)
+        bytes_consumed += sizeof buffer / 4;  // discard excessive garbage
```

Comment:
It is indeed arbitrary, as I don't have any good rule to tell what exact amount has to be discarded. As it stands, if we have an incomplete token in the buffer (i.e. not followed by white space) larger than 2&nbsp;KiB, we discard 1&nbsp;KiB.

Do you have a suggestion?

Reply:
@edgar-bonet there are at least two prior approaches for comparison to this new approach (let's call it "(c)"):
- a) `master` and release 1.5.2 would skip anything malformed until they find a line feed (`\n`)
- b) `development` since merge of pull request #98 would truncate the buffer down to all-empty once it is truly full.

Personally I like (b) best: it only acts when absolutely necessary but then gives the most relief possible, which may matter with high pressure input coming in at stdin. Both (b) and (c) have the potential to cut through the middle of a single token that would turn out a well-formed double with more bytes of input from the future later. Does (b) have some quality that you wanted to avoid when going for (c) or that motivate (c) in particular?

Reply:
There is one issue with approach (b). Imagine we get a very long token like `xxx...xy42` on stdin. The buffer gets full right before reading `y`. It gets discarded. On the next read, we try to parse `y42`. It is recognized as an invalid token. This is good. Now, imagine the buffer gets full right _after_ reading `y`. On the next read, we parse `42` as a valid token. This is bad.

The rationale of approach (c) is:

* It ensures that, after processing some input, there is always 2K of free space left in the buffer. This ensures that `read` is never called with a small `count` parameter, which would be bad in case of high input pressure.

* It makes very hard for a partial invalid token to be mistakenly parsed as a valid one.

Yes, there are still edge cases that could be mishandled, like a 2K-long `12.000...000` token. I do not see any easy way of handling this king of token, and I would not get of of my way to accomodate them.

Reply:
> The rationale of approach (c) is:
> 
> * It ensures that, after processing some input, there is always 2K of free space left in the buffer. This ensures that `read` is never called with a small `count` parameter, which would be bad in case of high input pressure.

@edgar-bonet so rather than requesting [0..4k] bytes per `read` call we would request [1k..4k] bytes now, interesting. My vote for briefly documenting that motivation as a comment in code. Why only discard `sizeof(buffer) / 4` rather than full `sizeof(buffer) / 2` when air is getting thin? If there is no particular reason, `/ 2` seems more natural and would not need an additional comment. It would be [2k..4k] for all requests then, which would support the case of high pressure.

> There is one issue with approach (b). Imagine we get a very long token like `xxx...xy42` on stdin. The buffer gets full right before reading `y`. It gets discarded. On the next read, we try to parse `y42`. It is recognized as an invalid token. This is good. Now, imagine the buffer gets full right _after_ reading `y`. On the next read, we parse `42` as a valid token. This is bad.
> 
> The rationale of approach (c) is:
> 
> * [..]
> 
>  * It makes very hard for a partial invalid token to be mistakenly parsed as a valid one.

The cutting still happens by position rather than content, the chances of cutting right though `y42` should be the same — am I missing something?

> Yes, there are still edge cases that could be mishandled, like a 2K-long `12.000...000` token. I do not see any easy way of handling this king of token, and I would not get of of my way to accomodate them.

We'd need a dynamically growing buffer for that. I'm with you that we don't need that complexity. PS: If current 4096 seems too low, I think we'd be free to pick 8192 also.

Reply:
> Why only discard `sizeof(buffer) / 4` rather than full `sizeof(buffer) / 2` when air is getting thin?

In order to avoid turning `xxx...xy42` into `42`.

> the chances of cutting right though y42 should be the same

There would be 1K right after the `42`, within the same token. The chances of this token being valid are slim.

Reply:
> My vote for briefly documenting that motivation as a comment in code.

Too tired to come up with something. Any suggestion?

Reply:
> > My vote for briefly documenting that motivation as a comment in code.
> 
> Too tired to come up with something. Any suggestion?

@edgar-bonet "Ensure that we can always ask `read` for >=2k bytes and not have that number go down near zero, for better performance, especially with high pressure input?" Something like that?

Reply:
> > Why only discard `sizeof(buffer) / 4` rather than full `sizeof(buffer) / 2` when air is getting thin?
> 
> In order to avoid turning `xxx...xy42` into `42`.
> 
> > the chances of cutting right though y42 should be the same
> 
> There would be 1K right after the `42`, within the same token. The chances of this token being valid are slim.

@edgar-bonet I think I understand now: at the moment where…

> `buffer_pos - bytes_consumed > sizeof(buffer) / 2`

…is checked, we know that all these bytes failed to make even a single valid token — which probably means that there is no whitespace in there anywhere, correct? — and so we cut off 1k of a block of 2k that only later could have become a >=2k single token. Does that sound right? Then this approach seems to indeed offer a bonus. Definitely worth commenting in code for the next generation on this code base.

Reply:
Thanks! Added a slight variation on that. Would you mind rebasing?

Reply:
@edgar-bonet I have pushed a rebased copy into branch `refactor-input-rebased` in here right now. Only the first commit had conflicts to be resolved. Please check if everything looks like expected and if so please free to replace the pull request content by `refactor-input-rebased` via reset hard and a forced pushed.

Reply:
@edgar-bonet the CI likes it: https://github.com/tenox7/ttyplot/actions?query=branch%3Arefactor-input-rebased

Reply:
Thank-you ! That was helpful. I took your rebased branch, restored `erase()` and `refresh()`, squashed a few commits together, and pushed again.